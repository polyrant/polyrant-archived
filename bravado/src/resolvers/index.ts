export { PostResolver } from './post';
export { PingResolver } from './ping';
export { UserResolver } from './user';
