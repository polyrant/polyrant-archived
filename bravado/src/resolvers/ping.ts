export default {
  Query: {
    ping: async (): Promise<string> => {
      return 'Pong!';
    },
  },
};
