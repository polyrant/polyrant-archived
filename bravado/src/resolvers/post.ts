import { Arg, Ctx, Int, Mutation, Query, Resolver } from 'type-graphql';

import { Post } from '../entities';
import { MyContext } from '../types';

@Resolver()
export class PostResolver {
  @Query(() => [Post])
  posts(@Ctx() { em }: MyContext): Promise<Post[]> {
    return em.find(Post, {});
  }

  @Query(() => Post, { nullable: true })
  post(
    @Arg('id', () => Int) id: number,
    @Ctx() { em }: MyContext
  ): Promise<Post | null> {
    return em.findOne(Post, { id });
  }

  @Mutation(() => Post)
  async createPost(
    @Arg('content') content: string,
    @Ctx() { em }: MyContext
  ): Promise<Post> {
    const post = em.create(Post, { content });
    await em.persistAndFlush(post);
    return post;
  }

  @Mutation(() => Boolean)
  async deletePost(
    @Arg('id') id: number,
    @Ctx() { em }: MyContext
  ): Promise<boolean> {
    try {
      em.nativeDelete(Post, { id });
      return true;
    } catch (err) {
      return false;
    }
  }
}
