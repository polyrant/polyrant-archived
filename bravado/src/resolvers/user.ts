import { IResolvers } from 'apollo-server-express';
import argon2 from 'argon2';

import { User } from '../entities';

import { MyContext, UserInput, UserResponse } from '../types';

const resolvers: IResolvers = {
  /************************* Mutations *************************/

  Mutation: {
    // Register a user
    register: async (
      _: void,
      input: UserInput,
      { em }: MyContext
    ): Promise<UserResponse> => {
      if (input.username.trim().length < 3)
        return {
          errors: [
            {
              field: 'Username',
              message: 'Username must be at least 3 characters!',
            },
          ],
        };

      if (input.password.length < 8)
        return {
          errors: [
            {
              field: 'Password',
              message: 'Password must be at least 8 characters!',
            },
          ],
        };

      const hashedPass = await argon2.hash(input.password);
      const user = em.create(User, {
        username: input.username,
        password: hashedPass,
      });

      try {
        await em.persistAndFlush(user);
      } catch (err) {
        if (err.code === '23505') {
          return {
            errors: [
              {
                field: 'Username',
                message: 'Username already exists!',
              },
            ],
          };
        }
      }

      return {
        user,
      };
    },

    // Log in a user
    login: async (
      _: void,
      input: UserInput,
      { em }: MyContext
    ): Promise<UserResponse> => {
      const user = await em.findOne(User, {
        username: input.username.toLowerCase(),
      });

      if (!user)
        return {
          errors: [
            {
              field: 'User',
              message: 'Incorrect Login Information!',
            },
          ],
        };

      const validPass = await argon2.verify(user.password, input.password);

      if (!validPass)
        return {
          errors: [
            {
              field: 'User',
              message: 'Incorrect Login Information!',
            },
          ],
        };

      return {
        user,
      };
    },
  },

  /*************************  *************************/
};

export default resolvers;
