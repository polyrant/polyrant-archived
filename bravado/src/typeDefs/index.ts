import { gql } from 'apollo-server-express';

export default gql`
  input UserInput {
    username: String!
    password: String!
  }

  type FieldError {
    field: String
    message: String
  }

  type UserResponse {
    errors: [FieldError]!
    user: User
  }

  type User {
    id: Int!
    createdAt: String!
    updatedAt: String!
    username: String!
    password: String!
    name: String
    bio: String
  }

  type Post {
    id: String!
    content: String!
    createdAt: String!
  }

  type Query {
    ping: String!

    posts: [Post]!
    post: Post
  }

  type Mutation {
    createPost: Post!
    deletePost: Boolean!

    register: UserResponse!
    login: UserResponse!
  }
`;
