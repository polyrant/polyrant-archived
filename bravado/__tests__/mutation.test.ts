import { ApolloServer } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';

import resolvers from '../src/resolvers';
import typeDefs from '../src/typeDefs';

const CREATE_POST = gql`
  mutation {
    createPost(content: "This is a post!") {
      content
    }
  }
`;

const DELETE_POST = gql`
  mutation {
    deletePost(id: 1)
  }
`;

const REGISTER = gql`
  mutation {
    register(input: { username: "test", password: "12345678" }) {
      errors {
        field
        message
      }

      user {
        username
        bio
        name
        updatedAt
        createdAt
      }
    }
  }
`;

const LOGIN = gql`
  mutation {
    login(input: { username: "test", password: "12345678" }) {
      user {
        username
        bio
        name
        createdAt
        updatedAt
      }
      errors {
        message
      }
    }
  }
`;

const server = async () => {
  const server = new ApolloServer({
    resolvers,
    typeDefs,
    mockEntireSchema: true,
    mocks: true,
  });

  return server;
};

describe('mutations', () => {
  test('createPost', async () => {
    const { mutate } = createTestClient(await server());

    const res = await mutate({ mutation: CREATE_POST });
    expect(res).toMatchSnapshot();
  });

  test('deletePost', async () => {
    const { mutate } = createTestClient(await server());

    const res = await mutate({ mutation: DELETE_POST });
    expect(typeof res.data.deletePost === 'boolean').toBeTruthy();
  });

  test('register', async () => {
    const { mutate } = createTestClient(await server());

    const res = await mutate({ mutation: REGISTER });
    expect(res).toMatchSnapshot();
  });

  test('login', async () => {
    const { mutate } = createTestClient(await server());

    const res = await mutate({ mutation: LOGIN });
    expect(res).toMatchSnapshot();
  });
});
