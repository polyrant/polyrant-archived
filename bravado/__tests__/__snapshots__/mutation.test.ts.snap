// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutations createPost 1`] = `
Object {
  "data": Object {
    "createPost": Object {
      "content": "Hello World",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`mutations login 1`] = `
Object {
  "data": Object {
    "login": Object {
      "errors": Array [
        Object {
          "message": "Hello World",
        },
        Object {
          "message": "Hello World",
        },
      ],
      "user": Object {
        "bio": "Hello World",
        "createdAt": "Hello World",
        "name": "Hello World",
        "updatedAt": "Hello World",
        "username": "Hello World",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`mutations register 1`] = `
Object {
  "data": Object {
    "register": Object {
      "errors": Array [
        Object {
          "field": "Hello World",
          "message": "Hello World",
        },
        Object {
          "field": "Hello World",
          "message": "Hello World",
        },
      ],
      "user": Object {
        "bio": "Hello World",
        "createdAt": "Hello World",
        "name": "Hello World",
        "updatedAt": "Hello World",
        "username": "Hello World",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
